lib/rspec/mocks/message_expectation.rb

```diff
@@ -85,6 +85,48 @@ def and_return(first_value, *values)
         nil
       end
 
+      # Tells the object to invoke a Proc when it receives the message. Given
+      # more than one value, the result of the first Proc is returned the first
+      # time the message is received, the result of the second Proc is returned
+      # the next time, etc, etc.
+      #
+      # If the message is received more times than there are Procs, the result of
+      # the last Proc is returned for every subsequent call.
+      #
+      # @return [nil] No further chaining is supported after this.
+      # @example
+      #   allow(api).to receive(:get_foo).and_invoke(-> { raise ApiTimeout })
+      #   api.get_foo # => raises ApiTimeout
+      #   api.get_foo # => raises ApiTimeout
+      #
+      #   allow(api).to receive(:get_foo).and_invoke(-> { raise ApiTimeout }, -> { raise ApiTimeout }, -> { :a_foo })
+      #   api.get_foo # => raises ApiTimeout
+      #   api.get_foo # => rasies ApiTimeout
+      #   api.get_foo # => :a_foo
+      #   api.get_foo # => :a_foo
+      #   api.get_foo # => :a_foo
+      #   # etc
+      def and_invoke(first_proc, *procs)
+        raise_already_invoked_error_if_necessary(__method__)
+        if negative?
+          raise "`and_invoke` is not supported with negative message expectations"
+        end
+
+        if block_given?
+          raise ArgumentError, "Implementation blocks aren't supported with `and_invoke`"
+        end
+
+        procs.unshift(first_proc)
```

Comment:
👍 I clearly understand that you enforce passing at least a single argument and rely on Ruby.

The mutating `unshift` slightly concerned me, in the following scenario:
```ruby
procs = [-> { 1 }, -> { raise }]
allow(foo).to receive(:bar).and_invoke(*procs)
allow(foo).to receive(:baz).and_invoke(*procs)
```
but on a synthetic test, it seems that it has no effect on this user-defined `procs`, so all good.

Reply:
Its not a directly passed in array, as its essentially the same as `and_invoke(-> { 1 }, -> { raise })` its also identical to our `and_return` implementation (I squashed a different quibble I had due to that).